<a href="https://colab.research.google.com/github/sonudk/Introduction_to_AI_and_IoT-/blob/master/Dealing_With_Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Dealing With Data </center>
## <center> Linear Separatrix </center>
### <center> Prof. Laxmidhar Behera </center>
### <center> IIT Kanpur </center>

## <center> Obstacle Avoidance </center>
<table><tr>
    <td><figure><img src=https://static.generation-robots.com/12044-large_default/robot-mobile-pioneer-3-dx.jpg alt="my img"/><figcaption> <center><font size="+2">Pioneer 3dx</font></center>  </figcaption></figure>
    </td>
    <td> <img src="https://drive.google.com/uc?id=1FRBvRTZTNAp8OBxInOtr9mmP0tIhmmYc" style="" ></td>
    <td> <img src="https://drive.google.com/uc?id=1mP2l7nB3WLmBH3lkEUmbki-52wRqjc_w" style="" ></td>
</tr></table>

In [0]:
%%HTML
<style>
td {
  font-size: 30px
}
th {
  font-size: 30px
}
.rendered_html {
      font-size:30px;
}
.row {
  display: flex;
}

.column {
  flex: 50%;
  font-size : 20px
}

</style>

## <center>What is data?</center>
| Sonar S | $v_l$   |$v_r$|
|---------|---------|-----|
|60cm |10cm/sec |10cm/sec |
| 50cm |9cm/sec| 9cm/sec |
| 30cm | 7cm/sec | 4cm/sec |
|20cm | 4cm/sec | 0cm/sec|

Model I : $v_l = a_1S + a_0 $ <br>
Model II : $v_l = a_2S^2 + a_1S + a_0 $ <br>

**These models are linear in terms of unknown parameters.**




### <center>Finding the linear separatrix</center>
Given data and linear model, data can be represented in terms of the following expression - <br>
$Aw = y$ &nbsp; $A \in R^{nxm}$ &nbsp; $w \in R^m$ &nbsp; $y \in R^n$ <br><br>
There are three possibilities:
1. $n = m$ &nbsp; Exact Solution
2. $n>m$ &nbsp;   No Exact Solution; Least Square solution exists
3. $n<m$ &nbsp;   Many solutions exist

### <center>Least Square Solution </center> <br>
<center>
$Aw = y$ &nbsp; $A \in R^{nxm}$ &nbsp; $w \in R^m$ &nbsp; $y \in R^n$ <br><br>
$E = \frac{1}{2}\tilde{y}^T\tilde{y}$ &nbsp; &nbsp; &nbsp; $\tilde{y} = Aw - y$ <br><br>
$\frac{\partial{E}}{\partial{w}} = 0$ &nbsp; &nbsp; &nbsp; $(Aw - y)^TA = 0$ <br><br>
$w = (A^TA)^{-1}A^Ty$
</center>

| Sonar S | $v_l$   |$v_r$|
|---------|---------|-----|
|60cm |10cm/sec |10cm/sec |
| 50cm |9cm/sec| 9cm/sec |
| 30cm | 7cm/sec | 4cm/sec |
|20cm | 4cm/sec | 0cm/sec|

<br>
<div class="row">
  <div class="column" align = "center">Model I <br>
$0.6a_1 + a_0 = 10$<br>
$0.5a_1 + a_0 = 9$<br>
$03a_1 + a_0 = 7$<br><br>
      </div>
    
  <div class="column" align = "center">Model II<br>
$0.36a_2 + 0.6a_1 + a_0 = 10$ <br>
$0.25a_2 + 0.5a_1 + a_0 = 9$ <br>
$0.09a_2 + 0.3a_1 + a_0 = 7$ <br>
</div>
</div>
<br>
<center><font size="5">$w = (A^TA)^{-1}A^Ty$</font></center>
<br>

<div class="row">
<div class="column">
    <div class="column">
      \begin{align}
      w &= \begin{bmatrix}
          14\\
          1.9
         \end{bmatrix}
      \end{align}
    </div>
    <div class="column">
      \begin{align}
      y &= \begin{bmatrix}
          10.3\\
          8.9 \\
          6.1 \\
          4.7
         \end{bmatrix}\\
      E = 1.4
         \end{align}

  </div>
  </div>
  <div class="column">
    <div class = "column">
      \begin{align}
      w &= \begin{bmatrix}
          -33.33\\
          40\\
          -2.6
         \end{bmatrix}
    \end{align}
    </div>
   <div class="column">
    \begin{align}
      y &= \begin{bmatrix}
          9.8\\
          9.4\\
          6.6\\
          4.2
         \end{bmatrix} \\
      E = 0.4
    \end{align}
    </div>
  </div>
</div>


<img src="https://drive.google.com/uc?id=1bqZwiOleqtC0wfOdLK1ZDlJcPoiGpEtu">


### IRIS flower dataset
<table><tr>
    <td><img src="https://drive.google.com/uc?id=1iItH7Ptc0uu2HcgJunEOnlDyOiJjpYgH" alt="iris_1"/></td>
    <td><img src="https://drive.google.com/uc?id=19TkNSuqGrSLWesTEKs1YHnYAWGRGTSBr" style="" ></td>
    <td><img src="https://drive.google.com/uc?id=1eh9bylEYc8dCZvalfcf-oNUePhScs4Sf" style="" ></td>
    <td><img src="https://drive.google.com/uc?id=1zwIWTyocOhlNUZ2-q6NZs9jLs8ucFAKY" style="" ></td>
</tr></table>
The data set consists of 50 samples from each of three species of Iris (Iris setosa, Iris virginica and Iris versicolor). <br>
Four features were measured from each sample: the length and the width of the sepals and petals, in centimeters.<br>
Based on the combination of these four features, Fisher developed a linear discriminant model to distinguish the species from each other. 

<img src="https://drive.google.com/uc?id=10ddnnmNyWNEu8V956LnGo5YA5FN39Hed" >

### <center>Loading Data in python</center>
Sklearn.datasets  is a library of many standard datasets including iris data <br>
Numpy is a numerical processing library in python <br>
Keras is a python library for building deep neural networks



In [0]:
# Import and load iris dataset from sklearn
from sklearn.datasets import load_iris
import numpy as np
import keras
np.random.seed(10)

iris = load_iris() # iris is a python dictionary with key-value pairs

Using TensorFlow backend.


##### Data Exploration

In [0]:
print('keys in iris dictionary: ', iris.keys())

X = iris['data']
print('First 3 entries of X:', '\n', X[:3])

Y = iris['target']
print('First 3 entries of Y:', '\n', Y[:3])

names = iris['target_names']
print('names:', names)
feature_names = iris['feature_names']
print('feature_names:', feature_names)

# Track a few sample points
isamples = np.random.randint(len(Y), size = (5))
print(isamples)

keys in iris dictionary:  dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])
First 3 entries of X: 
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]]
First 3 entries of Y: 
 [0 0 0]
names: ['setosa' 'versicolor' 'virginica']
feature_names: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
[  9 125  15  64 113]


In [0]:
# Shape of Data
print('Shape of X:', X.shape)
print('Shape of Y:', Y.shape)
print('X - samples:', X[isamples])
print('Y- samples:', Y[isamples])

Shape of X: (150, 4)
Shape of Y: (150,)
X - samples: [[4.9 3.1 1.5 0.1]
 [7.2 3.2 6.  1.8]
 [5.7 4.4 1.5 0.4]
 [5.6 2.9 3.6 1.3]
 [5.7 2.5 5.  2. ]]
Y- samples: [0 2 0 1 2]


#### Pre-process Data
Convert labels to categorical - One-hot encoding

In [0]:
from keras.utils import to_categorical

Ny = len(np.unique(Y)) # Ny is number of categories/classes
print('Ny: ', Ny)

Y = to_categorical(Y[:], num_classes = Ny) # converted to 1-hot

print('X - samples:', X[isamples])
print('Y- samples:', Y[isamples])

Ny:  3
X - samples: [[4.9 3.1 1.5 0.1]
 [7.2 3.2 6.  1.8]
 [5.7 4.4 1.5 0.4]
 [5.6 2.9 3.6 1.3]
 [5.7 2.5 5.  2. ]]
Y- samples: [[1. 0. 0.]
 [0. 0. 1.]
 [1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


#### Train - Test split


In [0]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.20, random_state = 1)
print('X_train shape:', X_train.shape)
print('X_test.shape:', X_test.shape)

X_train shape: (120, 4)
X_test.shape: (30, 4)


#### Data Normalization/Scaling
Normalize data to be of zero mean and unit variance


In [0]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train) # Computes the mean and standard deviation

X_train = scaler.transform(X_train) # Perform transformation: x = (x-mean)/std
X_test = scaler.transform(X_test)
print('X_train: \n', X_train[:5]) # first 5 samples of X_train
print('Y_train: \n', Y_train[:5])

X_train: 
 [[ 0.31553662 -0.04578885  0.44767531  0.23380268]
 [ 2.2449325  -0.04578885  1.29769171  1.39742892]
 [-0.2873996  -1.24028061  0.05100098 -0.15407273]
 [ 0.67729835 -0.52358555  1.01435291  1.13884531]
 [-0.04622511 -0.52358555  0.73101411  1.52672073]]
Y_train: 
 [[0. 1. 0.]
 [0. 0. 1.]
 [0. 1. 0.]
 [0. 0. 1.]
 [0. 0. 1.]]


<table>
  <tr>
    <th colspan="0">X</th>
      <th colspan="4">Y</th>
      <th colspan="6"></th>
      </tr> 
  <tr>
      <td>0.31</td><td>-0.04</td><td>0.45</td><td>0.23</td><td>0</td><td>1</td><td>0</td>
    </tr>
    <tr>
      <td>2.24</td><td>-0.46</td><td>1.30</td><td>1.40</td><td>0</td><td>0</td><td>1</td>
      </tr>
</table>
<br>
\begin{align}
l_sw_{00} + b_sw_{10} + l_pw_{20} + b_pw_{30} + w_{40} = y_0 \\
l_sw_{01} + b_sw_{11} + l_pw_{21} + b_pw_{31} + w_{41} = y_1 \\
l_sw_{02} + b_sw_{12} + l_pw_{22} + b_pw_{32} + w_{42} = y_2
\end{align}
<br>
\begin{align}
[l_s\: b_s\: l_p\: b_p\: 1] \begin{bmatrix}
          w_{00} &w_{01} &w_{02}\\
          w_{10} &w_{11} &w_{12}\\
          \vdots &\vdots &\vdots \\
          w_{40} &w_{41} &w_{42}
         \end{bmatrix} = [y_0\: y_1\: y_2]
\end{align}
<br><br>
$$
Aw = y\\
\implies w = (A_TA)^{-1}A^Ty
$$


#### Training with $Aw = Y$

In [0]:
# define function to add column of 1's
addlcol = lambda x: np.concatenate((x, np.ones((x.shape[0], 1))), axis = 1)

Ns, Nx = X_train.shape
print('Ns: ', Ns, 'Nx: ', Nx)

def find_weights(A, Y):
    print(A.shape)
    
    print(Y.shape)

    w = np.linalg.inv(A.T.dot(A)).dot(A.T.dot(Y))
    return w

A = addlcol(X_train)
Y = Y_train
w = find_weights(A, Y)


Ns:  120 Nx:  4
(120, 5)
(120, 3)


In [0]:
def evaluate(X, W, Yd, transform_X_a):
  a = transform_X_a(X)
  yd = np.argmax(Yd, axis = 1)
  y = np.argmax(a.dot(W), axis = 1)
  print('Confusion Matrix:')
  print(confusion_matrix(yd, y))


evaluate(X_train, w, Y_train, addlcol)
evaluate(X_test, w, Y_test, addlcol)

Confusion Matrix:
[[39  0  0]
 [ 0 22 15]
 [ 0  4 40]]
Confusion Matrix:
[[11  0  0]
 [ 0  6  7]
 [ 0  0  6]]


<table>
  <tr>
    <th colspan="0">X</th>
      <th colspan="4">Y</th>
      <th colspan="6"></th>
      </tr> 
  <tr>
      <td>0.31</td><td>-0.04</td><td>0.45</td><td>0.23</td><td>0</td><td>1</td><td>0</td>
    </tr>
    <tr>
      <td>2.24</td><td>-0.46</td><td>1.30</td><td>1.40</td><td>0</td><td>0</td><td>1</td>
      </tr>
</table>
\begin{align}
[l_s\: b_s\: l_p\: b_p\: l_s^2\: b_s^2\: l_p^2\: b_p^2\: 1] \begin{bmatrix}
          w_{00} &w_{01} &w_{02}\\
          w_{10} &w_{11} &w_{12}\\
          \vdots &\vdots &\vdots \\
          w_{80} &w_{81} &w_{82}
         \end{bmatrix} = [y_0\: y_1\: y_2]
\end{align}
<br><br>
$$
Aw = y\\
\implies w = (A_TA)^{-1}A^Ty
$$

In [0]:
addSqlcol = lambda x: np.concatenate((x, x**2, np.ones((x.shape[0], 1))), axis = 1)

A = addSqlcol(X_train)
Y = Y_train
w = find_weights(A, Y)
evaluate(X_train, w, Y_train, addSqlcol)
evaluate(X_test, w, Y_test, addSqlcol)

(120, 9)
(120, 3)
Confusion Matrix:
[[39  0  0]
 [ 0 35  2]
 [ 0  2 42]]
Confusion Matrix:
[[11  0  0]
 [ 0 13  0]
 [ 0  0  6]]
